tf_encrypted/protocol/aby3/aby3.py

```diff
@@ -0,0 +1,3355 @@
+                x = self.define_constant(x, share_type=share_type)
+                y = self.define_public_tensor(y, share_type=share_type)
+                return x, y
+
+            if isinstance(y, ABY3Tensor):
+                x = self.define_constant(
+                    x,
+                    apply_scaling=y.is_scaled,
+                    share_type=share_type,
+                    factory=y.backing_dtype,
+                )
+                return x, y
+
+            raise TypeError(("Don't know how to lift "
+                             "{}, {}").format(type(x), type(y)))
+
+        if isinstance(x, tf.Tensor):
+
+            if isinstance(y, (np.ndarray, int, float)):
+                x = self.define_public_tensor(x, share_type=share_type)
+                y = self.define_constant(y, share_type=share_type)
+                return x, y
+
+            if isinstance(y, tf.Tensor):
+                x = self.define_public_tensor(x, share_type=share_type)
+                y = self.define_public_tensor(y, share_type=share_type)
+                return x, y
+
+            if isinstance(y, ABY3Tensor):
+                x = self.define_public_tensor(
+                    x,
+                    apply_scaling=y.is_scaled,
+                    share_type=share_type,
+                    factory=y.backing_dtype,
+                )
+                return x, y
+
+            raise TypeError(("Don't know how to lift "
+                             "{}, {}").format(type(x), type(y)))
+
+        if isinstance(x, ABY3Tensor):
+
+            if isinstance(y, (np.ndarray, int, float)):
+                y = self.define_constant(
+                    y,
+                    apply_scaling=x.is_scaled,
+                    share_type=share_type,
+                    factory=x.backing_dtype,
+                )
+                return x, y
+
+            if isinstance(y, tf.Tensor):
+                y = self.define_public_tensor(
+                    y,
+                    apply_scaling=x.is_scaled,
+                    share_type=share_type,
+                    factory=x.backing_dtype,
+                )
+                return x, y
+
+            if isinstance(y, ABY3Tensor):
+                return x, y
+
+        raise TypeError(("Don't know how to lift "
+                         "{}, {}").format(type(x), type(y)))
+
+    @memoize
+    def add_n(self, tensors):
+        # TODO(Morten) we could optimize by doing lazy reductions, potentially
+        #              segmenting as needed
+        return reduce(lambda x, y: x + y, tensors)
+
+    @memoize
+    def sub(self, x, y):
+        x, y = self.lift(x, y)
+        return self.dispatch("sub", x, y)
+
+    @memoize
+    def negative(self, x):
+        x = self.lift(x)
+        return self.dispatch("negative", x)
+
+    @memoize
+    def mul(self, x, y):
+        x, y = self.lift(x, y)
+        return self.dispatch("mul", x, y)
+
+    @memoize
+    def mul2(self, x, y):
```

Comment:
also could use a docstring and/or more descriptive name

Reply:
Indeed they are two different truncation protocols in the paper. Perhaps it's best to remove `mul2` in the current version, because it is only more efficient than `mul` when precomputation can be properly executed. But I still do not have a best idea about how to implement this precomputation in TFE.

Reply:
> when precomputation can be properly executed. But I still do not have a best idea about how to implement this precomputation in TFE.

Would you mind elaborating on this?

Reply:
> > when precomputation can be properly executed. But I still do not have a best idea about how to implement this precomputation in TFE.
> 
> Would you mind elaborating on this?

In Figure 1 of the trunc2 protocol in the ABY3 paper, the algorithm is separated into two parts: "Preprocess" and "Online". `mul2` is only better than `mul` when ALL the "Preprocess" of ALL the multiplications can be executed together at the beginning. But this could make the library difficult to use: how can we anticipate ALL the multiplications, considering various algorithms, iterations, shapes of tensors, etc...

Reply:
> how can we anticipate ALL the multiplications, considering various ML algorithms, iterations, shapes of tensors, etc...

Thanks for the details!

I have something in mind for this so suggest we keep the code around.